Support Hadoop batch ingestion for druid-azure-extensions

spyk opened the pull request:
Add support for Hadoop batch ingestion in Azure blob storage. Fixes #5181.

**Note** Hadoop paths through wasb:// do not support colons in filenames. The colons were removed by using `ISODateTimeFormat.basicDateTime()` for segment paths, and are also explicitely replaced by underscores. So, older segments created with real-time ingestion will not be recognized any more and will have to be re-imported since the naming of segments has changed.

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPuller.java:
```diff
@@ -58,15 +60,27 @@ public AzureDataSegmentPuller(
       prepareOutDir(outDir);
 
-      final ByteSource byteSource = new AzureByteSource(azureStorage, containerName, blobPath);
+      log.info("Loading container: " + containerName + " - " + "blobPath:" + blobPath + " - outdir: " + outDir);
```
Use String.format syntax instead of string concats

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPuller.java:
```diff
@@ -58,15 +60,27 @@ public AzureDataSegmentPuller(
-      final ByteSource byteSource = new AzureByteSource(azureStorage, containerName, blobPath);
+      log.info("Loading container: " + containerName + " - " + "blobPath:" + blobPath + " - outdir: " + outDir);
+
+      Boolean blobPathIsHadoop = blobPath.contains(HADOOP_BLOB_PATH_IDENTIFIER);
```
Use `boolean`

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
   public String getPathForHadoop()
   {
-    return null;
+    String hadoopPath = String.format(
```
Use `StringUtils.format()`

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+        config.getAccount()
+    );
+
+    log.info("Using Azure blob storage Hadoop path: " + hadoopPath);
```
Use String.format syntax instead of concat

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+  @Override
+  public String getStorageDir(DataSegment dataSegment)
+  {
+
```
Extra line

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+
+    // Replace colons with underscores, since they are not supported through wasb:// prefix
+    return seg.replace(":", "_");
+
```
Extra line

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+        dataSegment.getShardSpec().getPartitionNum()
+    );
+
+    log.info("DataSegment: " + seg);
```
Use String.format syntax

leventov commented on extensions-contrib/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java:
```diff
@@ -158,4 +158,15 @@ public void uploadDataSegmentTest() throws StorageException, IOException, URISyn
+    String hadoopPath = pusher.getPathForHadoop();
+
+    Assert.assertEquals("wasb://container@account.blob.core.windows.net/", hadoopPath);
+
```
Extra line

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+
+    String seg = JOINER.join(
+        dataSegment.getDataSource(),
+        String.format(
```
Use `StringUtils.format()`

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
+    log.info("DataSegment: " + seg);
+
+    // Replace colons with underscores, since they are not supported through wasb:// prefix
+    return seg.replace(":", "_");
```
Why doing this replace? From where colons could appear in `seg`?

spyk replied:
The dataSegment.getVersion() string is actually in the form of '2018-01-05T14:54:09.295Z' in my setup, coming from Druid I guess. I moved the conversion there. There is also a dataSegment.getBinaryVersion() field, I renamed any such occurrences to 'binaryVersion' to avoid confusion.

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -72,7 +74,41 @@ public String getPathForHadoop(String dataSource)
-    return null;
+    String hadoopPath = String.format(
+        Locale.ENGLISH,
+        "%s://%s@%s.blob.core.windows.net/",
```
Please use `HADOOP_BLOB_PATH_IDENTIFIER` from `AzureDataSegmentPuller`

spyk replied:
done

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPuller.java:
```diff
@@ -38,6 +38,8 @@
 {
   private static final Logger log = new Logger(AzureDataSegmentPuller.class);
 
+  String HADOOP_BLOB_PATH_IDENTIFIER = "blob.core.windows.net";
```
`static final`

leventov replied:
From where does this constant come? Is it specified somewhere in Azure docs? Please add a Javadoc comment

spyk commented:
Thank you for the review @leventov . I implemented most of your comments I believe, let me know if you spot anything. These are the changes in detail (also in git commit):

* Rename version to binaryVersion where appropriate to avoid confusion
* Set default protocol to wasbs://, as recommended by the Azure docs
* Add link to Azure documentation for wasb(s):// path
* Remove any colons from the dataSegment.getVersion()
* Added test for dataSegment.getVersion colon is replaced
* Use StringUtils.format for String concatenation
* remove empty lines

jon-wei commented:
@leventov Do you have additional comments on this?

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPuller.java:
```diff
@@ -58,15 +66,29 @@ public AzureDataSegmentPuller(
       prepareOutDir(outDir);
 
-      final ByteSource byteSource = new AzureByteSource(azureStorage, containerName, blobPath);
+      log.info(StringUtils.format(
```
Don't need `StringUtils.format(` here, `log.info()` could accept exactly the same args.

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -70,7 +71,38 @@ public String getPathForHadoop(String dataSource)
+        AzureDataSegmentPuller.AZURE_STORAGE_HOST_ADDRESS
+    );
+
+    log.info(StringUtils.format("Using Azure blob storage Hadoop path: %s", hadoopPath));
```
Same

leventov commented on extensions-contrib/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java:
```diff
@@ -70,7 +71,38 @@ public String getPathForHadoop(String dataSource)
+        dataSegment.getShardSpec().getPartitionNum()
+    );
+
+    log.info(StringUtils.format("DataSegment: [%s]", seg));
```
Same

spyk commented:
Hi @leventov , updated the code with your comments (removed StringUtils.format from log.info). Let me know if you need anything else, thanks!

hoesler commented:
@spyk Thanks for adding this feature! I tried to get this working with a local mapred job but got an `java.lang.ClassNotFoundException: Class org.apache.hadoop.fs.azure.NativeAzureFileSystem not found`. The `hadoop-azure` lib is missing. I wonder if it should be added to druid-hdfs-storage, similar to [hadoop-aws](https://github.com/apache/incubator-druid/blob/04ea3c9f8c1f5ea34b023217bd709509ace4d30d/extensions-core/hdfs-storage/pom.xml#L148).

hoesler commented:
After adding it manually to druid-hdfs-storage, the class can be found. Indexing fails, however, with the following exception:

```
2018-07-16T16:35:11,102 ERROR [pool-33-thread-1] io.druid.indexer.JobHelper - Exception in retry loop
java.io.IOException: Cannot create file wasbs://*****@*****.blob.core.windows.net/wikiticker/2015-09-12T00:00:00.000Z_2015-09-13T00:00:00.000Z/2018-07-16T16:33:38.634Z/0/index.zip.0 through WASB that has colons in the name
```

Seems like hadoop-azure cannot deal with colons in the path: https://github.com/apache/hadoop/blob/b3fe56402d908019d99af1f1f4fc65cb1d1436a2/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/NativeAzureFileSystem.java#L1512

aryaflow commented:
Thanks for the feature. I tried this in a cluster and succeeded to make it work... but I had to add wasb to JobHelper.java as well.
`} else if ("wasbs".equals(type)) {
      segmentLocURI = URI.create(loadSpec.get("path").toString());
}`

aryaflow commented:
@hoesler hadoop-azure is not added by default. But it's needed.
